File: cp1Endsem/b.cpp
#include<bits/stdc++.h>

using namespace std;

#define ll long long
using namespace std;
using matrix = vector<vector<long long>>;
const long long MOD = 1e8 + 7;


matrix multiply(const matrix &a, const matrix &b) {
    ll n = a.size();
    matrix c(n, vector<long long>(n));
    for (ll i = 0; i < n; i++) {
        for (ll j = 0; j < n; j++) {
            for (ll k = 0; k < n; k++) {
                c[i][j] = (c[i][j] + a[i][k] * b[k][j]) % MOD;
            }
        }
    }
    return c;
}


matrix power(const matrix &base, long long exp) {
    if (exp == 1) {
        return base;
    }
    matrix half = power(base, exp / 2);
    matrix result = multiply(half, half);
    if (exp % 2 == 1) {
        result = multiply(result, base);
    }
    return result;
}

int main() {
    ll t;
    cin >> t;
    while (t--) {
        long long k;
        cin >> k;
        
        
        matrix trans = {{0, 1, 1, 1}, 
                        {1, 0, 1, 1}, 
                        {1, 1, 0, 1}, 
                        {1, 1, 1, 0}};

        
        matrix result = power(trans, k);

        
        cout << result[0][0] << endl;
    }
    return 0;
}

File: cp1Endsem/b.py
MOD = int(1e8 + 7)

def multiply(a, b):
    n = len(a)
    c = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                c[i][j] = (c[i][j] + a[i][k] * b[k][j]) % MOD
    return c

def power(base, exp):
    if exp == 1:
        return base
    half = power(base, exp // 2)
    result = multiply(half, half)
    if exp % 2 == 1:
        result = multiply(result, base)
    return result

def main():
    import sys
    input = sys.stdin.read
    data = input().split()
    
    t = int(data[0])
    index = 1
    results = []
    
    for _ in range(t):
        k = int(data[index])
        index += 1
        
        trans = [
            [0, 1, 1, 1], 
            [1, 0, 1, 1], 
            [1, 1, 0, 1], 
            [1, 1, 1, 0]
        ]
        
        result = power(trans, k)
        results.append(result[0][0])
    
    for res in results:
        print(res)

if _name_ == "_main_":
    main()